pkg/app/api/grpcapi/web_api.go

```diff
@@ -1311,6 +1311,98 @@ func (a *WebAPI) ListAPIKeys(ctx context.Context, req *webservice.ListAPIKeysReq
 }
 
 // GetInsightData returns the accumulated insight data.
-func (a *WebAPI) GetInsightData(_ context.Context, _ *webservice.GetInsightDataRequest) (*webservice.GetInsightDataResponse, error) {
+func (a *WebAPI) GetInsightData(ctx context.Context, req *webservice.GetInsightDataRequest) (*webservice.GetInsightDataResponse, error) {
+	claims, err := rpcauth.ExtractClaims(ctx)
+	if err != nil {
+		a.logger.Error("failed to authenticate the current user", zap.Error(err))
+		return nil, err
+	}
+
+	switch req.MetricsKind {
+	case model.InsightMetricsKind_DEPLOYMENT_FREQUENCY:
+		return a.getInsightDataForDeployFrequency(ctx, claims.Role.ProjectId, req)
+	}
 	return nil, status.Error(codes.Unimplemented, "")
 }
+
+// getInsightDataForDeployFrequency returns the accumulated insight data for deploy frequency.
+func (a *WebAPI) getInsightDataForDeployFrequency(ctx context.Context, projectID string, req *webservice.GetInsightDataRequest) (*webservice.GetInsightDataResponse, error) {
+	counts := make([]*model.InsightDataPoint, req.DataPointCount)
+
+	var movePoint func(time.Time, int) time.Time
+	var accumulateFrom time.Time
+	switch req.Step {
+	case model.InsightStep_DAILY:
+		movePoint = func(from time.Time, i int) time.Time {
+			return from.AddDate(0, 0, i)
+		}
+		accumulateFrom = time.Unix(req.RangeFrom, 0)
+	case model.InsightStep_WEEKLY:
+		movePoint = func(from time.Time, i int) time.Time {
+			return from.AddDate(0, 0, 7*i)
+		}
+		rangeFrom := time.Unix(req.RangeFrom, 0)
+		// Sunday in the week of rangeFrom
+		accumulateFrom = rangeFrom.AddDate(0, 0, -int(rangeFrom.Weekday()))
+	case model.InsightStep_MONTHLY:
+		movePoint = func(from time.Time, i int) time.Time {
+			return from.AddDate(0, i, 0)
+		}
+		rangeFrom := time.Unix(req.RangeFrom, 0)
+		accumulateFrom = time.Date(rangeFrom.Year(), rangeFrom.Month(), 1, 0, 0, 0, 0, time.Local)
+	case model.InsightStep_YEARLY:
+		movePoint = func(from time.Time, i int) time.Time {
+			return from.AddDate(i, 0, 0)
+		}
+		rangeFrom := time.Unix(req.RangeFrom, 0)
+		accumulateFrom = time.Date(rangeFrom.Year(), 1, 1, 0, 0, 0, 0, time.Local)
```

Comment:
The step is validated before sent, but let's handle the `default` case just in case.